src/object/srv_ec_aggregate.c

```diff
@@ -0,0 +1,1207 @@
+						(struct ec_agg_stripe_ud *)arg;
+	struct ec_agg_entry		*entry = stripe_ud->asu_agg_entry;
+	uint8_t				*bit_map = stripe_ud->asu_bit_map;
+	unsigned int			 cell_cnt = stripe_ud->asu_cell_cnt;
+	int				 rc = 0;
+
+	rc = agg_fetch_odata_cells(entry, bit_map, cell_cnt);
+	if (rc)
+		goto out;
+	rc = agg_update_parity(entry, bit_map, cell_cnt);
+
+out:
+	ABT_eventual_set(stripe_ud->asu_eventual, (void *)&rc, sizeof(rc));
+
+}
+
+/* Driver function for partial stripe update. Fetches the data and then invokes
+ * second function to update the parity.
+ */
+static int
+agg_process_partial_stripe(struct ec_agg_entry *entry)
+{
+	struct ec_agg_stripe_ud	 stripe_ud;
+	struct ec_agg_extent	*agg_extent;
+	unsigned int		len = entry->ae_oca->u.ec.e_len;
+	unsigned int		k = entry->ae_oca->u.ec.e_k;
+	unsigned int		i, cell_cnt = 0;
+	uint8_t			bit_map[OBJ_TGT_BITMAP_LEN] = {0};
+	int			*status;
+	int			rc = 0;
+
+	for (i = 0; i < k && cell_cnt < k; i++)
+		d_list_for_each_entry(agg_extent,
+				      &entry->ae_cur_stripe.as_dextents,
+				      ae_link) {
+			if (agg_overlap(&agg_extent->ae_recx, i, k, len,
+					entry->ae_cur_stripe.as_stripenum)) {
+				setbit(bit_map, i);
+				cell_cnt++;
+			}
+			if (cell_cnt == k)
+				break;
+		}
+
+
+	rc = agg_fetch_local_extents(entry, bit_map, cell_cnt);
+	if (rc)
+		goto out;
+
+	stripe_ud.asu_agg_entry = entry;
+	stripe_ud.asu_bit_map = bit_map;
+	stripe_ud.asu_cell_cnt = cell_cnt;
+	rc = ABT_eventual_create(sizeof(*status), &stripe_ud.asu_eventual);
+	if (rc != ABT_SUCCESS) {
+		rc = dss_abterr2der(rc);
+		goto out;
+	}
+	rc = dss_ult_create(agg_process_partial_stripe_ult, &stripe_ud,
+			    DSS_ULT_EC, 0, 0, NULL);
+	if (rc)
+		goto ev_out;
+	rc = ABT_eventual_wait(stripe_ud.asu_eventual, (void **)&status);
+	if (rc != ABT_SUCCESS) {
+		rc = dss_abterr2der(rc);
+		goto ev_out;
+	}
+	if (*status != 0) {
+		rc = *status;
+		goto ev_out;
+	}
+
+ev_out:
+	ABT_eventual_free(&stripe_ud.asu_eventual);
+
+out:
+	return rc;
+}
+
+/* Process the prior stripe. Invoked when the iterator has moved to the first
+ * extent in the subsequent.
+ */
+static int
+agg_process_stripe(struct ec_agg_entry *entry)
+{
+	vos_iter_param_t	iter_param = { 0 };
+	struct vos_iter_anchors	anchors = { 0 };
+	bool			update_vos = true;
+	int			rc = 0;
+
+	entry->ae_par_extent.ape_epoch	= ~(0ULL);
+
+	iter_param.ip_hdl		= DAOS_HDL_INVAL;
+	iter_param.ip_ih		= entry->ae_thdl;
+	iter_param.ip_flags		= VOS_IT_RECX_VISIBLE;
+	iter_param.ip_recx.rx_idx	= PARITY_INDICATOR |
+					  (entry->ae_cur_stripe.as_stripenum *
+						entry->ae_oca->u.ec.e_len);
+	iter_param.ip_recx.rx_nr	= entry->ae_oca->u.ec.e_len;
+
+	rc = vos_iterate(&iter_param, VOS_ITER_RECX, false, &anchors,
+			 agg_recx_iter_pre_cb, NULL, entry, NULL);
+	if (rc != 0)
+		goto out;
+
+	if (entry->ae_par_extent.ape_epoch > entry->ae_cur_stripe.as_hi_epoch &&
+			entry->ae_par_extent.ape_epoch != ~(0ULL)) {
+		/* Parity newer than data; nothing to do. */
```

Comment:
Shouldn't the data being deleted?

Reply:
You are correct, and this change has been made.